### ISSUE TITLE
Latest toil no longer works with cactus

### ISSUE THREAD

**Issue by glennhickey** [DataBiosphere/toil#2854]
A cactus command that works in version 3.20 now gives this error with master (I'm using a newer toil to test kubernetes for cactus).  It's strange because `Job._runner()` has a `defer` argument in the [source](https://github.com/DataBiosphere/toil/blob/master/src/toil/job.py#L1371-L1375):
```
INFO:toil.worker:Redirecting logging to /data1/cactus/toil-6141c5d4-202b-4550-a160-4d43177be980-565d6e0ff41b452899bd1cc744672e86/tmpj9fnaA/worker_log.txt
Job ended: 'RunCactusPreprocessorThenProgressiveDown' kind-RunCactusPreprocessorThenProgressiveDown/instanceICGaXj
The job seems to have left a log file, indicating failure: 'RunCactusPreprocessorThenProgressiveDown' kind-RunCactusPreprocessorThenProgressiveDown/instanceICGaXj
INFO:toil:Running Toil version 3.21.0a1-81a543d0e1a8e6c299f22bb8e862d34097e1f0bc.
WARNING:toil.resource:'JTRES_612c0b9862883c5800042ea7ac702778' may exist, but is not yet referenced by the worker (KeyError from os.environ[]).
Traceback (most recent call last):
  File "/data1/cactus/toil-venv-master/local/lib/python2.7/site-packages/toil/worker.py", line 366, in workerScript
    job._runner(jobGraph=jobGraph, jobStore=jobStore, fileStore=fileStore, defer=defer)
TypeError: _runner() got an unexpected keyword argument 'defer'
ERROR:toil.worker:Exiting the worker because of a failed job on host ip-172-31-26-21
```

This can be reproduced in a minute with these commands:
```
deactivate
git clone https://github.com/DataBiosphere/toil.git toil.master
git clone https://github.com/ComparativeGenomicsToolkit/cactus.git --recursive
cd cactus
virtualenv toil-venv-master
. toil-venv-master/bin/activate
pip install ../toil.master[all]
pip install .
toil clean jobStore ; cactus jobStore examples/evolverMammals.txt examples/evolverMammals.hal  --binariesMode docker --batchSystem singleMachine --workDir .
```

Doing the same thing with the last release runs okay
```
# Works with toil 3.20
deactivate
git clone https://github.com/DataBiosphere/toil.git toil.320
pushd toil.320 ;  git checkout cf34ca3416697f2abc816b2538f20ee29ba16932 ; popd
git clone https://github.com/ComparativeGenomicsToolkit/cactus.git --recursive
cd cactus
virtualenv toil-venv-320
. toil-venv-320/bin/activate
pip install ../toil.320[all]
pip install .
toil clean jobStore ; cactus jobStore examples/evolverMammals.txt examples/evolverMammals.hal  --binariesMode docker --batchSystem singleMachine --workDir .
```

┆Issue is synchronized with this [Jira Task](https://ucsc-cgl.atlassian.net/browse/TOIL-451)
┆Issue Number: TOIL-451

**Comment by glennhickey** [comment]
Oops, I probably should have posted the `--logDebug` error.  Here it is
```
DEBUG:toil.common:Obtained node ID 565d6e0ff41b452899bd1cc744672e86 from file /var/lib/dbus/machine-id
INFO:toil.worker:Redirecting logging to /data1/cactus/toil-28e47869-81d4-4053-afaa-a8e6100efc5a-565d6e0ff41b452899bd1cc744672e86/tmpw1C2zI/worker_log.txt
DEBUG:toil.deferred:Deleting /data1/cactus/toil-28e47869-81d4-4053-afaa-a8e6100efc5a-565d6e0ff41b452899bd1cc744672e86/deferred/tmpA24JkM
Finished job. self.coreFractions ~ 160 and self.memory ~ 128834523136
Ran jobID: 5 with exit value: 0
Job ended: 'RunCactusPreprocessorThenProgressiveDown' kind-RunCactusPreprocessorThenProgressiveDown/instanceTfugoW
Job 'RunCactusPreprocessorThenProgressiveDown' kind-RunCactusPreprocessorThenProgressiveDown/instanceTfugoW continues to exist (i.e. has more to do)
The job seems to have left a log file, indicating failure: 'RunCactusPreprocessorThenProgressiveDown' kind-RunCactusPreprocessorThenProgressiveDown/instanceTfugoW
kind-RunCactusPreprocessorThenProgressiveDown/instanceTfugoW    INFO:toil.worker:---TOIL WORKER OUTPUT LOG---
kind-RunCactusPreprocessorThenProgressiveDown/instanceTfugoW    INFO:toil:Running Toil version 3.21.0a1-81a543d0e1a8e6c299f22bb8e862d34097e1f0bc.
kind-RunCactusPreprocessorThenProgressiveDown/instanceTfugoW    DEBUG:toil:Configuration: {'maxLocalJobs': 16, 'rescueJobsFrequency': 3600, 'readGlobalFileMutableByDefault': False, 'logLevel': 'DEBUG', 'minNodes': None, 'targetTime': 1800, 'jobStore': u'file:/data1/cactus/jobStore', 'linkImports': False, 'manualMemArgs': False, 'maxJobDuration': 9223372036854775807, 'nodeOptions': None, 'nodeTypes': [], 'servicePollingInterval': 60, 'workDir': '/data1/cactus', 'stats': False, 'disableCaching': True, 'noStdOutErr': False, 'maxPreemptableServiceJobs': 9223372036854775807, 'environment': {}, 'parasolMaxBatches': 10000, 'cleanWorkDir': 'always', 'disableChaining': True, 'maxCores': 9223372036854775807, 'sseKey': None, 'maxMemory': 9223372036854775807, 'forceDockerAppliance': False, 'cwl': False, 'scaleInterval': 60, 'deadlockWait': 3600, 'defaultPreemptable': False, 'disableWorkerOutputCapture': False, 'clusterStats': None, 'defaultCores': 1, 'cseKey': None, 'betaInertia': 0.1, 'metrics': False, 'maxNodes': [10], 'scale': 1, 'writeLogs': None, 'runCwlInternalJobsOnWorkers': False, 'disableAutoDeployment': False, 'badWorker': 0.0, 'defaultDisk': 2147483648, 'mesosMasterAddress': '172.31.26.21:5050', 'restart': False, 'useAsync': True, 'preemptableCompensation': 0.0, 'parasolCommand': 'parasol', 'workflowID': '28e47869-81d4-4053-afaa-a8e6100efc5a', 'maxServiceJobs': 9223372036854775807, 'writeLogsFromAllJobs': False, 'badWorkerFailInterval': 0.01, 'statePollingWait': 1, 'debugWorker': False, 'maxLogFileSize': 64000, 'defaultMemory': 2147483648, 'workflowAttemptNumber': 0, 'maxDisk': 9223372036854775807, 'clean': 'onSuccess', 'provisioner': None, 'batchSystem': 'singleMachine', 'retryCount': 5, 'writeLogsGzip': None, 'nodeStorage': 50}
kind-RunCactusPreprocessorThenProgressiveDown/instanceTfugoW    DEBUG:toil.deferred:Running for file /data1/cactus/toil-28e47869-81d4-4053-afaa-a8e6100efc5a-565d6e0ff41b452899bd1cc744672e86/deferred/tmpA24JkM
kind-RunCactusPreprocessorThenProgressiveDown/instanceTfugoW    DEBUG:toil.worker:Parsed job wrapper
kind-RunCactusPreprocessorThenProgressiveDown/instanceTfugoW    DEBUG:toil.worker:Got a command to run: _toil files/for-job/kind-RunCactusPreprocessorThenProgressiveDown/instanceTfugoW/cleanup/file-mIPP8w/stream /data1/cactus/toil-venv-master/local/lib/python2.7/site-packages cactus.progressive.cactus_progressive True
kind-RunCactusPreprocessorThenProgressiveDown/instanceTfugoW    DEBUG:toil.job:Loading user module ModuleDescriptor(dirPath=u'/data1/cactus/toil-venv-master/local/lib/python2.7/site-packages', name=u'cactus.progressive.cactus_progressive', fromVirtualEnv=True).
kind-RunCactusPreprocessorThenProgressiveDown/instanceTfugoW    WARNING:toil.resource:'JTRES_612c0b9862883c5800042ea7ac702778' may exist, but is not yet referenced by the worker (KeyError from os.environ[]).
kind-RunCactusPreprocessorThenProgressiveDown/instanceTfugoW    DEBUG:toil.deferred:Running orphaned deferred functions
kind-RunCactusPreprocessorThenProgressiveDown/instanceTfugoW    DEBUG:toil.deferred:Running job
kind-RunCactusPreprocessorThenProgressiveDown/instanceTfugoW    DEBUG:toil.fileStores.abstractFileStore:LOG-TO-MASTER: Job files/for-job/kind-RunCactusPreprocessorThenProgressiveDown/instanceTfugoW/cleanup/file-mIPP8w/stream used 0.00% (8.0 KB [8192B] used, 2.0 GB [2147483648B] requested) at the end of its run.
kind-RunCactusPreprocessorThenProgressiveDown/instanceTfugoW    DEBUG:toil.deferred:Running own deferred functions
kind-RunCactusPreprocessorThenProgressiveDown/instanceTfugoW    DEBUG:toil.deferred:Out of deferred functions!
kind-RunCactusPreprocessorThenProgressiveDown/instanceTfugoW    DEBUG:toil.deferred:Running orphaned deferred functions
kind-RunCactusPreprocessorThenProgressiveDown/instanceTfugoW    Traceback (most recent call last):
kind-RunCactusPreprocessorThenProgressiveDown/instanceTfugoW      File "/data1/cactus/toil-venv-master/local/lib/python2.7/site-packages/toil/worker.py", line 366, in workerScript
kind-RunCactusPreprocessorThenProgressiveDown/instanceTfugoW        job._runner(jobGraph=jobGraph, jobStore=jobStore, fileStore=fileStore, defer=defer)
kind-RunCactusPreprocessorThenProgressiveDown/instanceTfugoW    TypeError: _runner() got an unexpected keyword argument 'defer'
kind-RunCactusPreprocessorThenProgressiveDown/instanceTfugoW    ERROR:toil.worker:Exiting the worker because of a failed job on host ip-172-31-26-21
kind-RunCactusPreprocessorThenProgressiveDown/instanceTfugoW    WARNING:toil.jobGraph:Due to failure we are reducing the remaining retry count of job 'RunCactusPreprocessorThenProgressiveDown' kind-RunCactusPreprocessorThenProgressiveDown/instanceTfugoW with ID kind-RunCactusPreprocessorThenProgressiveDown/instanceTfugoW to 0
Added job: 'RunCactusPreprocessorThenProgressiveDown' kind-RunCactusPreprocessorThenProgressiveDown/instanceTfugoW to active jobs
Built the jobs list, currently have 1 jobs to update and 0 jobs issued
Updating status of job 'RunCactusPreprocessorThenProgressiveDown' kind-RunCactusPrepr
```

**Comment by DailyDreaming** [comment]
@glennhickey I ran your commands and it's in one of the submodules that cactus pulls in.
If you run `git clone https://github.com/ComparativeGenomicsToolkit/cactus.git --recursive` it looks like it creates a `cactus/src/cactus/shared/common.py` file with an outdated `_runner()`.

**Comment by glennhickey** [comment]
Thanks @DailyDreaming.  Changing that version of `_runner` to accept and pass along `defer` fixes that error.  But it now crashes shortly after with this error about the `fileStore` going in not having the correct type.  Any ideas greatly appreciated!  Perhaps there's something else going on in cactus that needs porting to the new Toil interfaces?   

```
DEBUG:toil.deferred:Running own deferred functions
DEBUG:toil.deferred:Out of deferred functions!
DEBUG:toil.deferred:Running orphaned deferred functions
Traceback (most recent call last):
  File "/home/hickey/dev/cactus/local-venv/lib/python2.7/site-packages/toil/worker.py", line 367, in workerScript
    job._runner(jobGraph, jobStore, fileStore, defer)
  File "/home/hickey/dev/cactus/local-venv/lib/python2.7/site-packages/cactus/shared/common.py", line 1112, in _runner
    super(RoundedJob, self)._runner(jobGraph=jobGraph, jobStore=jobStore, fileStore=fileStore, defer=defer)
  File "/home/hickey/dev/cactus/local-venv/lib/python2.7/site-packages/toil/job.py", line 1392, in _runner
    returnValues = self._run(jobGraph, fileStore)
  File "/home/hickey/dev/cactus/local-venv/lib/python2.7/site-packages/toil/job.py", line 1329, in _run
    return self.run(fileStore)
  File "/home/hickey/dev/cactus/local-venv/lib/python2.7/site-packages/cactus/progressive/cactus_progressive.py", line 235, in run
    schedule.loadProject(self.project, fileStore=fileStore)
  File "/home/hickey/dev/cactus/local-venv/lib/python2.7/site-packages/cactus/progressive/schedule.py", line 74, in loadProject
    configFile = fileStore.readGlobalFile(exp.getConfigID())
  File "/home/hickey/dev/cactus/local-venv/lib/python2.7/site-packages/toil/fileStores/nonCachingFileStore.py", line 108, in readGlobalFile
    raise TypeError('Received file ID not of type FileID: {}'.format(fileStoreID))
TypeError: Received file ID not of type FileID: files/no-job/file-5YxEwM/cactus_config.xml
ERROR:toil.worker:Exiting the worker because of a failed job on host boreale
WARNING:toil.jobGraph:Due to failure we are reducing the remaining retry count of job 'RunCactusPreprocessorThenProgressiveDown' kind-RunCactusPreprocessorThenProgressiveDown/instancem6BaLw with ID kind-RunCactusPreprocessorThenProgressiveDown/instancem6BaLw to 0
Added job: 'RunCactusPreprocessorThenProgressiveDown' kind-RunCactusPreprocessorThenProgressiveDown/instancem6BaLw to active jobs
```

**Comment by joelarmstrong** [comment]
Oof, if toil now checks that files are FileID rather than string that will break things pretty severely. Is this related to the caching change? Do FileIDs now contain some extra data (in addition to the size)? For legacy reasons there is a lot of serialization/deserialization which would be a nightmare to track down (which is why all this XML stuff still exists). In fact up until a somewhat recent version some jobStores returned strings rather than FileIDs. (and fileIDs were just a superclass of str IIRC).

Can you use an earlier toil version?

**Comment by adamnovak** [comment]
This is related to the caching; we can't work out whether the cache can hold a file, or clear space for it, unless we know its size. I don't think our old behavior was really correct when we could handle non-FileIDs with the old caching implementation, so I added the check.

We could change the whole design to expose a way for each job store to tell you the size of a file, and go back to using real strings for file IDs, I guess. Then we would poll for the size internally instead of hoping the user code sent it along.

We can't use an older Toil here, unless we want to try backporting the Kubernetes support.

**Comment by glennhickey** [comment]
I've made some headway by making sure that all IDs in the experiment_wrapper XML structure go through `FileID.pack()` and `unpack()`, which allows size to be preserved in the string.  Now I get an error where 
```
snapshotExportID = job.fileStore.jobStore.getEmptyFileStoreID()
```
is just a string, which is hard to blame on cactus.

**Comment by adamnovak** [comment]
Hm. I think that in general you are Not Supposed to mix jobStore and fileStore calls like this.

What does Cactus think it is going to do with a FileID for an empty file? Files in the FileStore that you reference by ID can't be updated (only deleted); there's no way for an update at one node to be reflected by another node's cache. So if this ID is supposed to refer to a file in the FileStore that file will always be empty. But it looks to me like Cactus might be prepping to write some data at that ID later.

If you want a FileID for an empty file, you can get one from writeGlobalFileStream() and just write nothing; we could add a wrapper that does that, but I'm not sure about the general utility of getting a permanently empty file.

If you want to use the jobStore API, you have `updateFile`, but you don't have any caching, and tinkering around at that level with things the FileStore has stored there can cause Bad Things to happen. So part of the point of the FileID type is to prevent things from being brought from one of these APIs to another, without a lot of thought.

(`getEmptyFileStoreID` probably ought to be renamed, since it does sound like it might make a FileID, and since it really sits below the FileStore.)

In this case, if it worked before, you probably want to keep `snapshotExportID` as a string here, and make it a `FileID` when you actually fill it in (with `updateFile`/`updateFileStream` on the jobStore) and know its size.

**Comment by glennhickey** [comment]
I'm still getting weird errors deep in the run.  This fits with Joel's "breaking things pretty servlely" comment.  And I'm not sure it makes sense to wait on it for porting Cactus to kubernetes.  

@adamnovak can we disable the check for now in Toil.  Or have it only happen when caching is enabled in the file store?

**Comment by adamnovak** [comment]
@glennhickey We could disable the type checks in `nonCachingFileStore.py` (since the non-caching file store I don't think really *needs* the sizes) at least for the purpose of running Cactus. Do you want to do that, or do you want me to do it and prepare you a specially patched Docker?

I'm not sure we would want to merge that into Toil master, though. Those checks are there in the non-caching file store to make sure you don't write a workflow that suddenly fails when you go over to the caching file store.

We could also make job.fileStore.jobStore.getEmptyFileStoreID() actually return a FileID.

I think the Right Thing to do depends on whether we actually are going to change Cactus to use FileIDs eventually. If we are, the Right Thing to do is to temporarily disable the check for the non-caching file store, while we track down and fix all of the "weird errors deep in the run".

If we think changing how Cactus handles file IDs is too much work, then the Right Thing to do is get rid of the FileID class and its size field (or at least deprecate it) and add a `getFileSize(str)` method to the FileStore, backed by a similar method in the jobStore. Then file IDs go back to being bog-standard strings, and the caching logic queries the size instead of having it provided.

So shall we refactor Cactus or refactor Toil?

**Comment by glennhickey** [comment]
@adamnovak If you could make me a patched Docker, that'd be great (I have no idea how).  Thanks!  We want to get some monkeys aligning ASAP on kubernetes and this patch is looking like the only way forward for that. 

Long term's a harder question.  I guess I'll punt that to @benedictpaten to see what he wants to pay for.  Perhaps @joelarmstrong has some insight?

**Comment by adamnovak** [comment]
@glennhickey Try `TOIL_APPLIANCE_SELF=quay.io/adamnovak/toil:3.21.0a1-9a0a490a127756521d29dfc3baabc2dc672eac9a`; That's the https://github.com/DataBiosphere/toil/tree/issues/2854-hacktus branch I just made which should downgrade this error to a warning in the non-caching file store.

Isn't caching important for good performance in Cactus though?

**Comment by glennhickey** [comment]
Perfect, thanks!!

It's been too long since I've worked with Cactus for me to have much of an intuition of how much the file cache would help.  I know it'd be handy for toil-vg, where I want to have many jobs packing up the whole-genome GAM without needing to split it.

**Comment by adamnovak** [comment]
OK, I've written some code that should poll file sizes from the job store if you hand in something that doesn't have a `size`. Hopefully that fixes some of Cactus's problems. You will then be able to send just the string version of a FileID places or use a string from the jobStore if you really want to.

But as soon as you do something like download a file and then later modify it and then read it again, with caching enabled, you won't get any cache invalidation and your results will be *spooky*.

**Comment by adamnovak** [comment]
@glennhickey I made you another Docker: quay.io/adamnovak/toil:3.22.0a1-72750c175eb84fde0e6d3302283f032edc39038b

Does that one work for you? It stops warning and just gets the size from the job store if it needs it and doesn't have it. So you don't need to pack/unpack anything if you don't want to.

**Comment by glennhickey** [comment]
Thanks @adamnovak.  This version seems to be running okay (on a local cactus test, as well as toil-vg kubernetes).

**Comment by Secretloong** [comment]
Hi @adamnovak , I also met this problem when run cactus with `cactus --binariesMode local`. After updating the toil to version 3.22.0a1-91eab2b3a8c29e10db3a35a2ad5053fb462849c3, the cactus could be run without any ERROR reports. But it will repeat the failure and reissue cycles as bellow, it does not stop and there is no any results even no any progress for cactus pipeline.
```
Issuing the command: _toil_worker PreprocessSequence file:/data/cactus/examples/test_new_20191125/test_qsub kind-Pre
processSequence/instanceTB7QG4 with memory: 2097152000, cores: 1, disk: 104857600
Issued job 'PreprocessSequence' kind-PreprocessSequence/instanceTB7QG4 with job batch system ID: 17 and cores: 1, disk: 100.0 M, and memory: 2.0 G
Got exit code 1 (indicating failure) from job _toil_worker PreprocessSequence file:/data/cactus/examples/test_new_20191125/test_qsub kind-PreprocessSequence/instanceTB7QG4.
Job failed with exit value 1: 'PreprocessSequence' kind-PreprocessSequence/instanceTB7QG4
Job 'PreprocessSequence' kind-PreprocessSequence/instanceTB7QG4 continues to exist (i.e. has more to do)
No log file is present, despite job failing: 'PreprocessSequence' kind-PreprocessSequence/instanceTB7QG4
Due to failure we are reducing the remaining retry count of job 'PreprocessSequence' kind-PreprocessSequence/instanceTB7QG4 with ID kind-PreprocessSequence/instanceTB7QG4 to 5
We have increased the default memory of the failed job 'PreprocessSequence' kind-PreprocessSequence/instanceTB7QG4 to 2147483648 bytes
Added job: 'PreprocessSequence' kind-PreprocessSequence/instanceTB7QG4 to active jobs
Updating status of job 'PreprocessSequence' kind-PreprocessSequence/instanceTB7QG4 with ID kind-PreprocessSequence/instanceTB7QG4: with result status: 1
```
I have run this test successfully by cactus installed last year. But I can not go through the new cactus and new toil.
Could you give helps for this?
Thanks

**Comment by adamnovak** [comment]
```
No log file is present, despite job failing: 'PreprocessSequence' kind-PreprocessSequence/instanceTB7QG4
```

This makes your problem somewhat impossible to debug remotely @Secretloong.

@glennhickey is 3.22.0a1-91eab2b3a8c29e10db3a35a2ad5053fb462849c3 still the Toil version you are using successfully, or do you have a newer/better one? And what commit of Cactus are you using?

@Secretloong if imitating Glenn's Toil/Cactus version exactly doesn't make it start working, can you share your data  and your command line command so we can replicate your problem?

If not, the way to diagnose this would be to add a bunch of RealtimeLogging logging statements to the PreprocessSequence job's source code to report on what it is doing:

```
from toil.realtimeLogger import RealtimeLogger
...
RealtimeLogger.info('Doing something blah blah')
RealtimeLogger.info('Running command {}'.format(command))
```

Then you would run the workflow with the options `--logInfo --realTimeLogging` so you can see those log messages in the leader's log whether the job's local log manages to get collected or not. Then you would be able to figure out exactly where/how PreprocessSequence is failing, and thus what the bug in Toil and/or Cactus actually is.

**Comment by glennhickey** [comment]
I'd recommend just Toil 3.20.0 and the latest master branch of Cactus.

**Comment by Secretloong** [comment]
@glennhickey perfect thanks, Toil 3.20.0 works very well. 

@adamnovak the data is the example in Cactus software, and the command: 
```
cactus --binariesMode local --logFile test.log test_workplace/  evolverMammals.txt test.hal --disableCaching --disableHotDeployment --maxCores 4 --stats --workDir temp/
```
the new toil doesn't work with cactus in AWS too, I'm not sure is this solved now. I attach the relevant issue here #https://github.com/ComparativeGenomicsToolkit/cactus/issues/100#issuecomment-554526292

Thanks for all of you.
